rsky-relay/src/crawler/manager.rs

```diff
@@ -0,0 +1,151 @@
+use std::sync::atomic::Ordering;
+use std::time::Duration;
+use std::{io, thread};
+
+use magnetic::Consumer;
+use magnetic::buffer::dynamic::DynamicBufferP2;
+use sled::Tree;
+use thiserror::Error;
+
+use crate::SHUTDOWN;
+use crate::crawler::RequestCrawl;
+use crate::crawler::types::{Command, CommandSender, RequestCrawlReceiver, Status, StatusReceiver};
+use crate::crawler::worker::{Worker, WorkerError};
+use crate::types::{DB, MessageSender};
+
+const CAPACITY: usize = 1024;
+const SLEEP: Duration = Duration::from_millis(10);
+
+#[derive(Debug, Error)]
+pub enum ManagerError {
+    #[error("spawn error: {0}")]
+    Spawn(#[from] io::Error),
+    #[error("worker error: {0}")]
+    Worker(#[from] WorkerError),
+    #[error("rtrb error: {0}")]
+    Push(#[from] Box<rtrb::PushError<Command>>),
+    #[error("sled error: {0}")]
+    Sled(#[from] sled::Error),
+    #[error("join error")]
+    Join,
+}
+
+impl From<rtrb::PushError<Command>> for ManagerError {
+    fn from(value: rtrb::PushError<Command>) -> Self {
+        Box::new(value).into()
+    }
+}
+
+#[derive(Debug)]
+struct WorkerHandle {
+    pub command_tx: CommandSender,
+    pub thread_handle: thread::JoinHandle<Result<(), WorkerError>>,
+}
+
+pub struct Manager {
+    workers: Box<[WorkerHandle]>,
+    next_id: usize,
+    hosts: Tree,
+    request_crawl_rx: RequestCrawlReceiver,
+    status_rx: StatusReceiver,
+}
+
+impl Manager {
+    pub fn new(
+        n_workers: usize, message_tx: &MessageSender, request_crawl_rx: RequestCrawlReceiver,
+    ) -> Result<Self, ManagerError> {
+        #[expect(clippy::unwrap_used)]
+        let (status_tx, status_rx) =
+            magnetic::mpsc::mpsc_queue(DynamicBufferP2::new(CAPACITY).unwrap());
+        let workers = (0..n_workers)
+            .map(|worker_id| -> Result<_, ManagerError> {
+                let message_tx = message_tx.clone();
+                let status_tx = status_tx.clone();
+                let (command_tx, command_rx) = rtrb::RingBuffer::new(CAPACITY);
+                let thread_handle =
+                    thread::Builder::new().name(format!("rsky-crawl-{worker_id}")).spawn(
+                        move || Worker::new(worker_id, message_tx, command_rx, status_tx)?.run(),
+                    )?;
+                Ok(WorkerHandle { command_tx, thread_handle })
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        let hosts = DB.open_tree("hosts")?;
+        Ok(Self {
+            workers: workers.into_boxed_slice(),
+            next_id: 0,
+            hosts,
+            request_crawl_rx,
+            status_rx,
+        })
+    }
+
+    pub fn run(mut self) -> Result<(), ManagerError> {
+        for res in &self.hosts {
+            let (hostname, cursor) = res?;
+            let hostname = unsafe { String::from_utf8_unchecked(hostname.to_vec()) };
```

Comment:
Can we use [from_utf8](https://doc.rust-lang.org/std/str/fn.from_utf8.html) instead and add a Utf8Error error enum to `ManagerError`?

Reply:
there are unsafes like this elsewhere.